internal/cli/clients/client.go

```diff
@@ -1346,59 +1353,54 @@ func (c *EpinioClient) Push(ctx context.Context, name, rev, source string, param
 		log.V(3).Info("upload response", "response", upload)
 
 		gitRef = upload.Git
-	} else {
+	} else if params.GitRev != "" {
 		gitRef = &models.GitRef{
 			URL:      source,
-			Revision: rev,
+			Revision: params.GitRev,
 		}
 	}
 
-	c.ui.Normal().Msg("Staging application ...")
-
-	route, err := appDefaultRoute(ctx, appRef.Name)
-	if err != nil {
-		return errors.Wrap(err, "unable to determine default app route")
-	}
-	req := models.StageRequest{
-		App:       appRef,
-		Instances: params.Instances,
-		Git:       gitRef,
-		Route:     route,
-	}
-	details.Info("staging code", "Git", gitRef.Revision)
-	stage, err := c.stageCode(req)
-	if err != nil {
-		return err
-	}
-	log.V(3).Info("stage response", "response", stage)
-
-	details.Info("start tailing logs", "StageID", stage.Stage.ID)
+	stageID := ""
+	if params.Docker == "" {
+		c.ui.Normal().Msg("Staging application ...")
+		req := models.StageRequest{
+			App:   appRef,
+			Git:   gitRef,
+			Route: route,
+		}
+		details.Info("staging code", "Git", gitRef.Revision)
+		stage, err := c.stageCode(req)
+		if err != nil {
+			return err
+		}
+		stageID = stage.Stage.ID
+		log.V(3).Info("stage response", "response", stage)
 
-	// Buffered because the go routine may no longer be listening when we try
-	// to stop it. Stopping it should be a fire and forget. We have wg to wait
-	// for the routine to be gone.
-	stopChan := make(chan bool, 1)
-	var wg sync.WaitGroup
-	wg.Add(1)
-	defer wg.Wait()
-	go func() {
-		defer wg.Done()
-		err := c.AppLogs(appRef.Name, stage.Stage.ID, true, stopChan)
+		details.Info("start tailing logs", "StageID", stage.Stage.ID)
+		err = c.stageLogs(ctx, details, appRef, stage.Stage.ID)
 		if err != nil {
-			c.ui.Problem().Msg(fmt.Sprintf("failed to tail logs: %s", err.Error()))
+			return err
 		}
-	}()
+	}
 
-	details.Info("wait for pipelinerun", "StageID", stage.Stage.ID)
-	err = c.waitForPipelineRun(ctx, appRef, stage.Stage.ID)
+	c.ui.Normal().Msg("Deploying application ...")
+	deployRequest := models.DeployRequest{
+		App:            appRef,
+		Instances:      params.Instances,
+		Route:          route,
+		Git:            gitRef,
+		CustomImageURL: params.Docker,
+	}
+	if stageID != "" {
+		deployRequest.Stage = models.StageRef{ID: stageID}
```

Comment:
Should the deploy step care how the image was created? Can't it simply accept an `ImageURL` (replacing the `CustomImageURL`) every time? The internal registry is nothing but a convenience, maybe we don't need to handle it differently. We can keep the StageID around though. It is useful information for the Deployment and the rest of the resources.

Reply:
Right, I refactored it into using two variables ImageURL and RegistryURL. The imageURL is sent to the deploy endpoint when docker image is specified by the user else in all other cases, registryURL is sent to the deploy endpoint since we don't know ImageURL (the complete one). Savy?

Reply:
Done.